examples/basic_adding.py

```diff
@@ -24,162 +23,203 @@
+        distance = abs(ideal_price - provide_state["px"])
+        if distance > ALLOWABLE_DEVIATION * ideal_distance:
+            oid = provide_state["oid"]
+            print(f"Cancelling order due to deviation: oid:{oid}, side:{side}, ideal_price:{ideal_price}")
+            response = self.exchange.cancel(COIN, oid)
+            if response["status"] == "ok":
+                self.recently_cancelled_oid_to_time[oid] = get_timestamp_ms()
+                self.provide_state[side] = {"type": "cancelled"}
+            else:
+                logging.error(f"Failed to cancel order {oid} for side {side}: {response}")
+
+    def check_in_flight_order(self, side: Side, provide_state: InFlightOrder) -> None:
+        """Check if the in-flight order has timed out."""
+        if get_timestamp_ms() - provide_state["time"] > ORDER_TIMEOUT:
+            print("Order is still in flight after timeout, treating as cancelled.")
+            self.provide_state[side] = {"type": "cancelled"}
+
+    def place_new_order(self, side: Side, ideal_price: float) -> None:
+        """Place a new order if conditions are met."""
+        if self.position is None:
+            logging.debug("Waiting for position refresh before placing order.")
+            return
+
+        size = MAX_POSITION + self.position * side_to_int(side)
+        if size * ideal_price < 10:
+            logging.debug("Order size too small, not placing order.")
+            return
+
+        px = float(f"{ideal_price:.5g}")
+        print(f"Placing order: size:{size}, price:{px}, side:{side}")
+        response = self.exchange.order(COIN, side == "B", size, px, {"limit": {"tif": "Alo"}})
+        if response["status"] == "ok":
+            status = response["response"]["data"]["statuses"][0]
+            if "resting" in status:
+                self.provide_state[side] = {"type": "resting", "px": px, "oid": status["resting"]["oid"]}
 
     def on_user_events(self, user_events: UserEventsMsg) -> None:
+        """Callback for user events (e.g., fills)."""
         print(user_events)
-        user_events_data = user_events["data"]
-        if "fills" in user_events_data:
+        if "fills" in user_events["data"]:
             with open("fills", "a+") as f:
-                f.write(json.dumps(user_events_data["fills"]))
-                f.write("\n")
+                f.write(json.dumps(user_events["data"]["fills"]) + "\n")
         # Set the position to None so that we don't place more orders without knowing our position
         # You might want to also update provide_state to account for the fill. This could help avoid sending an
         # unneeded cancel or failing to send a new order to replace the filled order, but we skipped this logic
         # to make the example simpler
         self.position = None
 
-    def poll(self):
+    def poll(self) -> None:
+        """Poll open orders and user positions periodically."""
         while True:
+            time.sleep(POLL_INTERVAL)
```

Comment:
It's better to put this at the end of the loop so that the position is refreshed immediately when starting up